resolver_conn_wrapper.go

```diff
@@ -65,104 +70,116 @@ type ccResolverWrapperOpts struct {
 // newCCResolverWrapper uses the resolver.Builder to build a Resolver and
 // returns a ccResolverWrapper object which wraps the newly built resolver.
 func newCCResolverWrapper(cc resolverStateUpdater, opts ccResolverWrapperOpts) (*ccResolverWrapper, error) {
+	ctx, cancel := context.WithCancel(context.Background())
 	ccr := &ccResolverWrapper{
 		cc:                  cc,
-		done:                grpcsync.NewEvent(),
+		closed:              grpcsync.NewEvent(),
 		channelzID:          opts.channelzID,
 		ignoreServiceConfig: opts.bOpts.DisableServiceConfig,
+		outgoingCh:          buffer.NewUnbounded(),
+		serializer:          grpcsync.NewCallbackSerializer(ctx),
+		serializerCancel:    cancel,
 	}
 
-	var err error
-	// We need to hold the lock here while we assign to the ccr.resolver field
-	// to guard against a data race caused by the following code path,
-	// rb.Build-->ccr.ReportError-->ccr.poll-->ccr.resolveNow, would end up
-	// accessing ccr.resolver which is being assigned here.
-	ccr.resolverMu.Lock()
-	defer ccr.resolverMu.Unlock()
-	ccr.resolver, err = opts.builder.Build(opts.target, ccr, opts.bOpts)
+	r, err := opts.builder.Build(opts.target, ccr, opts.bOpts)
 	if err != nil {
+		cancel()
 		return nil, err
 	}
+	ccr.resolver = r
 	return ccr, nil
 }
 
 func (ccr *ccResolverWrapper) resolveNow(o resolver.ResolveNowOptions) {
-	ccr.resolverMu.Lock()
-	if !ccr.done.HasFired() {
+	ccr.serializer.Schedule(func(_ context.Context) {
 		ccr.resolver.ResolveNow(o)
-	}
-	ccr.resolverMu.Unlock()
+	})
 }
 
 func (ccr *ccResolverWrapper) close() {
-	ccr.resolverMu.Lock()
-	ccr.resolver.Close()
-	ccr.done.Fire()
-	ccr.resolverMu.Unlock()
+	ccr.closed.Fire()
+
+	done := make(chan struct{})
+	ccr.serializer.Schedule(func(_ context.Context) {
+		ccr.resolver.Close()
+		close(done)
+	})
+	<-done
+	ccr.serializerCancel()
```

Comment:
I think that would work and would be nice. But we need a little bit of help from the serializer.

Currently, the serializer accepts a context as part of its `New()` function and the only way to stop the serializer is to cancel the context. The `run()` method where the callbacks are processed in order has a `select` over the context's done channel and the other channel from which it reads callbacks. So, while cancelling the context does guarantee that no more callbacks are read from the `callbacks` channel, it does not provide a way to block on that event. See: https://github.com/grpc/grpc-go/blob/master/internal/grpcsync/callback_serializer.go#L55. So, I added a `Done` channel to the serializer. I felt that it might come in handy for other users as well.